### PR TITLE
Error text stays when user clicks away

### PR DESCRIPTION
**Pull request by fr0gs**
This PR fixes https://github.com/EFForg/privacybadger/issues/1965 . I am however unable to create a successful selenium test and emulate the "clicking away" action with the test framework. 

Refreshing the window or closing & reopening creates new **window** and thus destroying the previous state apparently

### CODE REVIEW

**Comment by ghostwords** [comment]
Sorry, apologies for the delay, but I haven't forgotten, on my list.

File: src/js/webrequest.js [hunk 1]
```diff
@@ -701,7 +701,14 @@ function dispatcher(request, sender, sendResponse) {
     let tab_id = request.tabId,
       tab_url = request.tabUrl,
       tab_host = window.extractHostFromURL(tab_url),
-      has_tab_data = badger.tabData.hasOwnProperty(tab_id);
+      has_tab_data = badger.tabData.hasOwnProperty(tab_id),
+      errorText = null;
```
Comment (ghostwords) [inline src/js/webrequest.js]: Nit: Non-object variables should use underscores: `error_text`.

File: src/js/webrequest.js [hunk 2]
```diff
@@ -776,6 +784,16 @@ function dispatcher(request, sender, sendResponse) {
     badger.storage.getBadgerStorageObject("action_map").deleteItem(request.origin);
 
     sendResponse();
+  } else if (request.type == "saveErrorText") {
+    let activeTab = badger.tabData[request.tabId];
```
Comment (ghostwords) [inline src/js/webrequest.js]: Could `badger.tabData[request.tabId]` ever be missing at this point?

File: src/js/webrequest.js [hunk 3]
```diff
@@ -776,6 +784,16 @@ function dispatcher(request, sender, sendResponse) {
...
+    activeTab.errorText = request.errorText;
+
+    sendResponse();
+  } else if (request.type == "removeErrorText") {
```
Comment (ghostwords) [inline src/js/webrequest.js]: Nit: Should add a blank line above.

File: src/js/webrequest.js [hunk 4]
```diff
@@ -776,6 +784,16 @@ function dispatcher(request, sender, sendResponse) {
...
+    let activeTab = badger.tabData[request.tabId];
+    delete activeTab.errorText;
+
+    // sendResponse();
```
Comment (ghostwords) [inline src/js/webrequest.js]: No need to leave this in.

File: src/js/webrequest.js [hunk 5]
```diff
@@ -776,6 +784,16 @@ function dispatcher(request, sender, sendResponse) {
     badger.storage.getBadgerStorageObject("action_map").deleteItem(request.origin);
 
     sendResponse();
+  } else if (request.type == "saveErrorText") {
+    let activeTab = badger.tabData[request.tabId];
+    activeTab.errorText = request.errorText;
+
+    sendResponse();
```
Comment (ghostwords) [inline src/js/webrequest.js]: Do we need the `sendResponse`? We don't do anything on response to this message back in the popup.

File: src/js/webrequest.js [hunk 6]
```diff
@@ -776,6 +784,16 @@ function dispatcher(request, sender, sendResponse) {
     badger.storage.getBadgerStorageObject("action_map").deleteItem(request.origin);
 
     sendResponse();
+  } else if (request.type == "saveErrorText") {
```
Comment (ghostwords) [inline src/js/webrequest.js]: Blank line above here too please.

File: src/js/popup.js [hunk 7]
```diff
@@ -110,6 +110,17 @@ function init() {
   var overlay = $('#overlay');
   $("#error").on("click", function() {
     overlay.toggleClass('active');
+
+    $('#error_input').bind('input propertychange', function() {
```
Comment (ghostwords) [inline src/js/popup.js]: Do we rebind over and over, once per error button click? We should move the binding then to do it only once.

File: src/js/popup.js [hunk 8]
```diff
@@ -346,6 +362,12 @@ function registerToggleHandlers() {
  * @param {Integer} tabId The id of the tab
  */
 function refreshPopup() {
+
+  //If there is any saved error text, fill the error input with it.
```
Comment (ghostwords) [inline src/js/popup.js]: Should probably move this paragraph down a bit, after the special/no tab data check if/else. Just for readability.

Would activate the overlay here (probably ...) too in order to reopen the popup straight into error reporting UI when the user left the popup with some error description text already entered.

File: src/js/webrequest.js (hunk 1 above)
Comment (fr0gs) [reply src/js/webrequest.js]: Done.

File: src/js/webrequest.js (hunk 2 above)
Comment (fr0gs) [reply src/js/webrequest.js]: tNo, the message comes from popup.js with the tabId set to `POPUP_DATA.tabId` attribute set at the beginning of it's execution.

```
$(function () {
  getTab(function (tab) {
    chrome.runtime.sendMessage({
      type: "getPopupData",
      tabId: tab.id,
      tabUrl: tab.url
    }, (response) => {
      setPopupData(response);
      refreshPopup();
      init();
    });
  });
});
```

I can only think of getTab failing and thus the `tab` variable not being passed in the callback, but that should be taken care of way before the popup messaging logic starts.

File: src/js/webrequest.js (hunk 3 above)
Comment (fr0gs) [reply src/js/webrequest.js]: Done.

File: src/js/webrequest.js (hunk 5 above)
Comment (fr0gs) [reply src/js/webrequest.js]: Certainly, it is more of a deformation of my coding, I tend to return anything always in a function.

**Comment by ghostwords** [comment]
Hi, just making sure you are not waiting on anything from me. Let me know when you want me to look this over again.

**Comment by fr0gs** [comment]
Hello, 

I don't have time to invest in this PR anymore, sorry. I hope the code may serve as a baseline for someone to take over.

**Comment by ghostwords** [comment]
OK, thanks for letting me know. This is pretty close to finished; perhaps we'll come back to it one day.

**Pull request closed without merging**